### DISCUSSION TITLE
Future developments

### DISCUSSION THREAD

**Discussion by apozas** [ecboghiu/inflation#92] [Ideas]
I suggest we store here the plans we have for the future. The idea is keeping track of the new features we want to implement, an estimation of its priority, and related information like concrete ideas for implementation.

# Short term goals
- Have a honest discussion on what we want to allow to do and what not with the library, and how we shall allow that (see #71 and #86).
- Add support for the user specifying arbitrary linear equality and inequality constraints. We have quite a bit of this already there (sanitization of monomials, processing to the solver...), so this should be relatively low effort.
- Add support for polynomial objectives. This can already be done by hand, but it will probably be useful to have it done automatically, so we can handle easily problems such as Eq. 26 in the [Quantum Inflation paper](https://arxiv.org/abs/1909.10519).
- Implement hybrid scenarios through an appropriate handling of `InflationSDP._notcomm`. This will require also of developing a way to allow the user to specify the nature of each of the sources. A first idea is specifying it with a list, with the order corresponding to that of `InflationProblem.dag.keys()` and where each element is `c` or `q`, that is fed to `InflationSDP` in a function that must be run before `generate_relaxation`.
- Using also an appropriate handling of `InflationSDP._notcomm`, add support for separable/entangling measurements.
- Create a parent class for `InflationSDP`, from where the current `InflationSDP` and a new `SupportsSDP` inherit. The motivation is that supports problem are different enough to standard problems (no ability to specify objectives, no processing of certificates) that it is risky and hard to not allow when dealing with supports all the things that you could do with a standard problem. 

# Medium term goals
- Support other solvers. The input of the new functions should be very similar to the current input to `solveSDP_mosekFUSION`.
- Allow to add new symmetries, like invariance under permutation of outcomes. In order to do this properly, we would want to develop abstract functions that take a permutation of the list of generating monomials and obtain the symmetrized moment matrix and orbits. In fact, this is exactly what we do in apply_inflation_symmetries, but we should think about it carefully. See #15.
- Improve the way of specifying DAGs. This is necessary if we want to allow for visible-to-latent connections and parentless visible-to-visible connections (see #51 for a bit more discussion). This last ones can be "hacked" in our code using `"dummy: ["visible_parent"], "visible_parent": ["visible_child"]`, but we should make it easier. Classical/quantum sources, and separable/entangling measurements may motivate adding properties to the nodes in the DAG and reading them off inside `InflationProblem`>

# Long term goals
- Implement an `InflationLP` class. This shall be general enough so as to be able to handle classical and GPT inflation. Then, a different subclass may be needed for each case. I am writing this in the long-term goal list because it is not just writing the code. This will involve writing tests, documentation, examples, etc.

**Comment by eliewolfe** [comment]
A reminder to split the support stuff to a subclass of InflationSDP. This class should have the ability to recover all probabilities from the moment matrix if it is feasible.

On a related note, we should check if all things would should factorize actually factorize. This applies both to support instances and standard distribution instances. Unknowns can factorize into unknowns, and do-conditionals can factorize (beyond networks). It would be nice to have a function to check if these _unenforcable_ factorization happen to hold, and to return to the user the set of all unenforceable factorization that a feasible SDP failed to enforce.

**Comment by eliewolfe** [comment]
To add to the short term list: a function `close_under_inflation_symmetries` which will expand the generating set as needed to make things nice. 

I'm going to move the LP version up to a slightly higher priority to _work on_ (but not to release it publicly) only because I am co-located with Victor Gitton at the moment and I think we can make use of that opportunity to make progress.
He is also going to focus on discovering symmetries of a given distribution, using using those to supplement the inflation symmetries.

**Comment by ecboghiu** [comment]
Something nice but not urgent (medium goal) would be to **parallelise the calculation of the moment matrix**. I believe I know how to do this, especially if using the "store (i,j) pairs" approach as opposed to having a dense moment matrix. 

Basically we can divide the moment matrix into several blocks and calculate the moments that appear in each block in parallel. We can modify calculate_momentmatrix to work exactly the same but to specify given a generating set what which i,j indices to loop through, as opposed to through all of them.  The sizes of the blocks should be determined such that all the different threads should take a similar time to compute, so that we don't have a bottleneck with one of the threads. Each thread would return a dictionary with keys hashes of the symmetrised monomial and value lists of pairs of (i,j) indices in the corresponding block where they appear. Merging together the results of the different threads is simply taking all the dictionaries of the different blocks and creating a single unique dictionary. This can be done by *concatenating* (i,j) indices of the variables that appear in several dictionaries. Identifies the overlappin variables should be very fast with set operations. Once we have this dictionary of (i,j) indices, initiating a dense moment matrix should be very fast, as we can use Pythonic language to assign its value momentmatrix[mon2ij[mon]] = index, where mon2ij[mon]= [[1,2],[2,1],...].

This whole process I expect to only give an advantage (if any) for large moment matrices. For large moment matrices computing mask matrices is very slow. This would helpy by: 1) creating the mask matrices from the beginning without the need for a conversion and 2) it would do it in a parallelised way. We would also not need to use scipy.sparse matrices until the very last moment, to initialise them from the lists of index pairs.

I will do this at some point and compare performance for large moment matrices.

**Comment by eliewolfe** [comment]
I've also been thinking about calculate_moment_matrix being a slight
bottleneck. I've had 2 thoughts about improving performance there.
1) Numba-fication. This is not so easy, as at the moment we use tobytes to
hash the 2d arrays, which Numba doesn't play nice with. We could use
array_equal loops, but that seems like it might be slower... My thought is
to create a "guvectorize" version of dot_mon which then acts like a NumPy
ufunc with automatic parallization, and which will then support the "outer"
method of ufuncs. The problem, however, is that guvectorize requires that
NumPy array returned by the guvectorized function must have the SAME
DIMENSION as one of the inputs (or rather, the dimensions of the output
must be a tensor with dimensions chosen from the among the dimensions of
the inputs). So, I think we'd need to change our data structure for a
monomial so that it's length is independent of the number of non-identity
operators in it. A headache, so one I'm inclined to not implement unless we
REALLY needed it.
2) Perform symmetrization BEFORE (and during) calculate moment matrix! This
is actually not so hard to implement. Basically, we use inflation and
transpose symmetries to get the POTENTIALLY unique matrix elements. We
would then apply dot_mon to the pairs of generating monomials (checking
against the reversal as well) to figure out if certain matrix elements
which are not NECESSARILY equivalent from the symmetries are nevertheless
equivalent due to commutation and removal of square projectors. I guess
this could work nicely in concert with Emi's proposal to break the
candidate (i,j) values into blocks for parallization. Note that instead of
selecting the minimal_by_hash version of just dot_mon and it's reverse, now
we would need to select minimal_by_hash taking into account all possible
sources relabellings as well.  I'm therefore skeptical if this will
actually be faster: we apply computions on fewer matrix elements, but the
computation on each element will be more involved, so hard to predict the
overall impact.

**Comment by eliewolfe** [comment]
I've been learning a lot about performance opportunities that have come up when working on InflationLP that can be transported back to InflationSDP.

1. Choice of data structure BEFORE initializing CompoundMonomial. In InflationSDP we were using a 2d numpy array to represent (ordered) lists of operators. But we can also simply use a 1d array representation, which records the indices of the operators in _lexorder. I've changed over InflationLP to use this 1d array representation in https://github.com/ecboghiu/inflation/commit/24faef0d4bd96fe0181ea680997d81c98351cd2b. We see significant reduction of memory overhead during the early processing, which counts for a lot, since symmetries and commutation rules eventually reduce the number of unique matrix elements. The hashing of monomials is also faster, since we use ndarray.tobytes() on small 1d arrays instead of 2d arrays. But far and away I expect the largest benefit will come in the actual moment matrix construction (which dots operator lists together). As of now, that operation involves converting the 2d to 1d, dotting, and then converting back. We should see a really noticeable improvement if we avoid the 2d representation until CompoundMonomial is initialized. Now, the use of 2d arrays should still be **acceptable** in the early stages (such as to specify custom NPA levels), but ultimately internally we would convert from 2d to 1d. I can probably complete this transition in a day or two of coding. It won't break functionality, but it will make the functions a bit more opaque. The in-docstring-examples would have the go (or be radically changes) for many functions. Most such functions are not really intended for the user anyway, so I think it is worth it, but we should have a chat before undertaking this transition.
2. Choice of data structure for representing equalities, inequalities, upper bounds, lowerbounds, semiknowns, objective, etc. Right now we are using **dictionaries** in InflationSDP. By contrast, in InflationLP, @chewtle has been helping me internally represent everything as coo_matrices (or coo_vectors). These sparse vectorized representations have lots of benefits, but primarily, they are faster to work with than dictionaries. All tests of solve, however, and the functions in optimization_utils, rely on dictionaries. We are intending to write a wrapper allowing solve to seamlessly worth with either dictionary representation or sparse matrix representation. This too could be emulated in InflationSDP. [Aside: the calculation of equality constraints in InflationSDP is significantly out of date relative to the approach in InflationLP. We should port the use of Boole-Frechet equalities instead of merely normalization...]
